File: djames-py/djames/linalg.py
"""Matrices over F_q: sampling, inversion, rank.

Matrices are held as lists of rows of small ints -- readable, and the natural
shape for the scheme's bookkeeping.  Only Gauss-Jordan runs on the packed
representation, because S and T reach 578 x 578 and an elimination written
over Python lists would be O(n^3) interpreter steps.

Packed elimination keeps lanes *unreduced* while sweeping a column: over odd
F_p a row accumulates at most n * (p-1)^2 before anything is read back, which
sits far below the 40-bit lanes used here, so no reduction is needed in the
inner loop.
"""

from .ff import F2Vec, F2kVec, FpVec
from .symmetric import sample_fq


def _packed(fq):
    if fq.q == 2:
        return F2Vec()
    if fq.p == 2:
        return F2kVec(fq)
    return FpVec(fq.q, 40)


def random_matrix(fq, nrows, ncols, xof):
    flat = sample_fq(xof, nrows * ncols, fq.q)
    return [flat[i * ncols:(i + 1) * ncols] for i in range(nrows)]


def identity(fq, n):
    return [[1 if i == j else 0 for j in range(n)] for i in range(n)]


def _echelon(fq, M, ncols, augment=None):
    """Reduced row echelon form of M (optionally carrying an augmentation).

    Returns (rank, rows, aug_rows) with rows/aug_rows as packed vectors.
    """
    V = _packed(fq)
    q = fq.q
    rows = [V.from_list(r) for r in M]
    aug = [V.from_list(r) for r in augment] if augment is not None else None
    nrows = len(rows)
    rank = 0
    for c in range(ncols):
        piv = None
        for i in range(rank, nrows):
            if V.coef(V.reduce(rows[i]), c) % q:
                piv = i
                break
        if piv is None:
            continue
        rows[rank], rows[piv] = rows[piv], rows[rank]
        if aug is not None:
            aug[rank], aug[piv] = aug[piv], aug[rank]
        rows[rank] = V.reduce(rows[rank])
        pv = V.coef(rows[rank], c) % q
        if pv != 1:
            s = fq.inv(pv)
            rows[rank] = V.reduce(V.scal(rows[rank], s))
            if aug is not None:
                aug[rank] = V.reduce(V.scal(aug[rank], s))
        if aug is not None:
            aug[rank] = V.reduce(aug[rank])
        for i in range(nrows):
            if i == rank:
                continue
            v = V.coef(rows[i], c) % q
            if v:
                f = fq.neg(v)
                rows[i] = V.add(rows[i], V.scal(rows[rank], f))
                if aug is not None:
                    aug[i] = V.add(aug[i], V.scal(aug[rank], f))
        rank += 1
        if rank == nrows:
            break
    rows = [V.reduce(r) for r in rows]
    if aug is not None:
        aug = [V.reduce(r) for r in aug]
    return rank, rows, aug, V


def rank(fq, M, ncols):
    return _echelon(fq, M, ncols)[0]


def invert(fq, M):
    """Inverse of a square matrix, or None if singular."""
    n = len(M)
    rk, rows, aug, V = _echelon(fq, M, n, augment=identity(fq, n))
    if rk != n:
        return None
    return [V.to_list(r, n) for r in aug]


def random_invertible(fq, n, xof):
    """A uniform invertible n x n matrix, together with its inverse.

    Rejection sampling: a uniform matrix over F_q is invertible with
    probability prod_{i=1..n} (1 - q^-i), about 0.29 for q = 2 and rising
    quickly with q, so the loop turns over a handful of times at most.
    """
    while True:
        M = random_matrix(fq, n, n, xof)
        Minv = invert(fq, M)
        if Minv is not None:
            return M, Minv


def random_full_rank(fq, nrows, ncols, xof):
    """A uniform nrows x ncols matrix of full row rank (nrows <= ncols)."""
    while True:
        M = random_matrix(fq, nrows, ncols, xof)
        if rank(fq, M, ncols) == nrows:
            return M


def mat_vec(fq, M, v):
    """M . v  (v a column of length ncols) -> list of length nrows."""
    out = []
    for row in M:
        acc = 0
        for c, x in zip(row, v):
            if c and x:
                acc = fq.add(acc, fq.mul(c, x))
        out.append(acc)
    return out


def vec_mat(fq, v, M):
    """v . M  (v a row of length nrows) -> list of length ncols."""
    ncols = len(M[0])
    out = [0] * ncols
    for x, row in zip(v, M):
        if x:
            for j, c in enumerate(row):
                if c:
                    out[j] = fq.add(out[j], fq.mul(x, c))
    return out

File: djames-rs/src/linalg.rs
//! Matrices over `F_q`: sampling, inversion, rank.
//!
//! Rows are `Vec<u8>`, one coefficient per byte, row-major.
//!
//! # Constant time
//!
//! Elimination runs a fixed nest of loops over every row and column and uses
//! masks rather than branches, so for a **full-rank square** input -- which is
//! what `S` and `T` always are -- the path taken is independent of the
//! entries. The rank counter advances on a mask, so a rank-deficient input
//! does take a different path; that only ever happens to a *candidate* matrix
//! that is then rejected and discarded, never to accepted key material.
//!
//! The one residual leak is `random_full_rank` for `M_Z` (`r x n`, `r = 2`):
//! which columns end up as pivots depends on the accepted matrix. For a
//! uniform matrix these are the first `r` columns with overwhelming
//! probability, so the leak is slight, but it is real and worth naming.

use crate::ct;
use crate::gf::Fq;
use crate::symmetric::{sample_fq, Xof};
use alloc::vec;
use alloc::vec::Vec;

pub type Matrix = Vec<Vec<u8>>;

/// A uniform `rows x cols` matrix, filled row-major from the stream.
pub fn random_matrix(xof: &mut Xof, rows: usize, cols: usize, q: u32) -> Matrix {
    let flat = sample_fq(xof, rows * cols, q);
    flat.chunks(cols).map(|c| c.to_vec()).collect()
}

pub fn identity(n: usize) -> Matrix {
    (0..n)
        .map(|i| (0..n).map(|j| u8::from(i == j)).collect())
        .collect()
}

/// Row-reduce `m` (carrying `aug` along), returning the rank.
fn echelon(fq: &Fq, m: &mut Matrix, ncols: usize, mut aug: Option<&mut Matrix>) -> usize {
    let nrows = m.len();
    let mut rank = 0usize;
    for c in 0..ncols {
        if rank >= nrows {
            break;
        }
        // Fold every lower row into the pivot row under the mask "the pivot
        // row's entry is still zero". After this the pivot is nonzero iff the
        // column has any nonzero entry at or below `rank`.
        for i in rank + 1..nrows {
            let need = ct::eq_mask32(m[rank][c] as u32, 0) as u8 & 1;
            for j in 0..ncols {
                let v = fq.mul(m[i][j], need);
                m[rank][j] = fq.add(m[rank][j], v);
            }
            if let Some(a) = aug.as_deref_mut() {
                for j in 0..a[i].len() {
                    let v = fq.mul(a[i][j], need);
                    a[rank][j] = fq.add(a[rank][j], v);
                }
            }
        }
        let piv = m[rank][c];
        if piv == 0 {
            continue; // rank-deficient: only reachable for rejected candidates
        }
        let inv = fq.inv(piv);
        for j in 0..ncols {
            m[rank][j] = fq.mul(m[rank][j], inv);
        }
        if let Some(a) = aug.as_deref_mut() {
            for j in 0..a[rank].len() {
                a[rank][j] = fq.mul(a[rank][j], inv);
            }
        }
        for i in 0..nrows {
            if i == rank {
                continue;
            }
            let f = fq.neg(m[i][c]);
            if f != 0 {
                for j in 0..ncols {
                    let v = fq.mul(m[rank][j], f);
                    m[i][j] = fq.add(m[i][j], v);
                }
                if let Some(a) = aug.as_deref_mut() {
                    for j in 0..a[rank].len() {
                        let v = fq.mul(a[rank][j], f);
                        a[i][j] = fq.add(a[i][j], v);
                    }
                }
            }
        }
        rank += 1;
    }
    rank
}

pub fn rank(fq: &Fq, m: &Matrix, ncols: usize) -> usize {
    let mut work = m.clone();
    echelon(fq, &mut work, ncols, None)
}

/// Inverse of a square matrix, or `None` if singular.
pub fn invert(fq: &Fq, m: &Matrix) -> Option<Matrix> {
    let n = m.len();
    let mut work = m.clone();
    let mut aug = identity(n);
    let rk = echelon(fq, &mut work, n, Some(&mut aug));
    if rk == n {
        Some(aug)
    } else {
        None
    }
}

/// A uniform invertible `n x n` matrix and its inverse.
///
/// Rejection is over the *sampled* matrix, so the retry count is independent
/// of the value finally returned. A uniform matrix over `F_q` is invertible
/// with probability `prod (1 - q^-i)`, about 0.29 for `q = 2` and rising with
/// `q`, so the loop turns over a handful of times at worst.
pub fn random_invertible(xof: &mut Xof, n: usize, q: u32) -> (Matrix, Matrix) {
    let fq = Fq::new(q);
    loop {
        let m = random_matrix(xof, n, n, q);
        if let Some(inv) = invert(&fq, &m) {
            return (m, inv);
        }
    }
}

/// A uniform `rows x cols` matrix of full row rank.
pub fn random_full_rank(xof: &mut Xof, rows: usize, cols: usize, q: u32) -> Matrix {
    let fq = Fq::new(q);
    loop {
        let m = random_matrix(xof, rows, cols, q);
        if rank(&fq, &m, cols) == rows {
            return m;
        }
    }
}

/// `M . v`, with `v` a column of length `cols`.
pub fn mat_vec(fq: &Fq, m: &Matrix, v: &[u8]) -> Vec<u8> {
    m.iter()
        .map(|row| {
            let mut acc = 0u8;
            for (&c, &x) in row.iter().zip(v.iter()) {
                acc = fq.add(acc, fq.mul(c, x));
            }
            acc
        })
        .collect()
}

/// `v . M`, with `v` a row of length `rows`.
pub fn vec_mat(fq: &Fq, v: &[u8], m: &Matrix) -> Vec<u8> {
    let cols = m[0].len();
    let mut out = vec![0u8; cols];
    for (&x, row) in v.iter().zip(m.iter()) {
        for j in 0..cols {
            out[j] = fq.add(out[j], fq.mul(x, row[j]));
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::symmetric::Xof;

    #[test]
    fn inverse_round_trips() {
        for q in [2u32, 4, 5, 13, 23] {
            let fq = Fq::new(q);
            let mut x = Xof::new(&[b"la", &[q as u8]]);
            let (m, mi) = random_invertible(&mut x, 12, q);
            for i in 0..12 {
                let e: Vec<u8> = (0..12).map(|k| u8::from(k == i)).collect();
                assert_eq!(vec_mat(&fq, &vec_mat(&fq, &e, &m), &mi), e, "q={q}");
                assert_eq!(vec_mat(&fq, &vec_mat(&fq, &e, &mi), &m), e, "q={q}");
            }
            let mut x = Xof::new(&[b"z", &[q as u8]]);
            let z = random_full_rank(&mut x, 3, 12, q);
            assert_eq!(rank(&fq, &z, 12), 3);
        }
    }
}
